### ISSUE TITLE
iobroker out node should create folder objects

### ISSUE THREAD

**Issue by mickym2** [ioBroker/ioBroker.node-red#206]
Currently an hierachie is created for non existing objects, if I specify a msg.topic in iobroker.out node.
The interim hierachies were correctly shown in admin 4 as folders if there are child objects. As admin 4 didn't check if an hierarchy level  is represented by an object - it was always possible to create any object type on any hierarchy level. 

With admin 5 you cannot create any child objects - as you have to create all parent objects manually. So it would be useful that all hierachies which are created by an iobroker out node, should be explicitely creates as folder objects.

**Comment by mickym2** [comment]
So if the iobroker out node does not create a folder object in case of errors the log will contain warnings like this:

`
<html>
<body>


ode-red.0 | 2021-08-10 22:02:18.991 | warn | This object will not be created in future versions. Please report this to the developer.
-- | -- | -- | --
node-red.0 | 2021-08-10 22:02:18.992 | warn | This object will not be created in future versions. Please report this to the developer.
node-red.0 | 2021-08-10 22:02:18.990 | warn | Object 0_userdata.0.controll-own.0.logic.states.fensterOffen is invalid: obj.common.type has an invalid value (undefined) but has to be one of number, string, boolean, array, object, mixed, file, json
node-red.0 | 2021-08-10 22:02:18.990 | warn | Object 0_userdata.0.controll-own.0.logic.states.fensterOffen.wohnzimmer is invalid: obj.common.type has an invalid value (undefined) but has to be one of number, string, boolean, array, object, mixed, file, json
node-red.0 | 2021-08-10 22:02:18.991 | warn | Object 0_userdata.0.controll-own.0.logic.states.fensterOffen.schlafzimmer is invalid: obj.common.type has an invalid value (undefined) but has to be one of number, string, boolean, array, object, mixed, file, json
node-red.0 | 2021-08-10 22:02:18.991 | warn | Object 0_userdata.0.controll-own.0.logic.states.fensterOffen.buero is invalid: obj.common.type has an invalid value (undefined) but has to be one of number, string, boolean, array, object, mixed, file, json
node-red.0 | 2021-08-10 22:02:18.990 | warn | This object will not be created in future versions. Please report this to the developer.
node-red.0 | 2021-08-10 22:02:18.989 | warn | This object will not be created in future versions. Please report this to the developer.
node-red.0 | 2021-08-10 22:02:18.989 | warn | This object will not be created in future versions. Please report this to the developer.
node-red.0 | 2021-08-10 22:02:18.990 | warn | This object will not be created in future versions. Please report this to the developer.
node-red.0 | 2021-08-10 22:02:18.988 | warn | Object 0_userdata.0.controll-own.0.logic.alarm.battery is invalid: obj.common.type has an invalid value (undefined) but has to be one of number, string, boolean, array, object, mixed, file, json


</body>
</html>`

**Comment by mickym2** [comment]
Perhaps it is possible that the iobroker system (or JS) creates automatically folder objects, if object type is undefined.

**Comment by mickym2** [comment]
OK - I was able to copy a tree, when iobroker out is configured to write "writeable" objects and with ACK-Flag. However the main problem is still the problem that folder objects have to be created. See the following screenshot with the source and the target tree created reading the objects with the objects list node and the iobroker out node

source = mqtt.1.*

![image](https://user-images.githubusercontent.com/58858526/129013551-ad003ea5-947d-4b8e-beff-c3e38ab60513.png)

target = 0_userdata.0.testcopy

![image](https://user-images.githubusercontent.com/58858526/129013718-4cc8402e-a652-4165-ac08-ac4ce08b6b90.png)

So even if the source has folder objects now - writing to the target path does not create folder objects on each level.

**Comment by mickym2** [comment]
I don't think so. As the iobroker out normally write states the problem are the missing folder objects in an id (as path), specified in msg.topic. The same happens if I specify an id for a state in admin. So if folder objects would be created automatically by the system, It isn't necessary that the iobroker out node creates explicitely objects or folders.

**Comment by Apollon77** [comment]
@mickym2 WHaat do you think, should the node automatically create the folder objects or should it be configurable?

**Comment by mickym2** [comment]
In my opinion it needs not to be configurable as in long term i guess it is intended to have on each hierarchy level an object and not how it is the case now. If a user doesn't want to create non existing states - this option is already available.
For point of view of NodeRed it should be completely transparent. 

The folder objects should be created automatically but should have the lowest priority. Means if later a state, device or channel object should be written with the iobroker out node to an existing folder object, this can be replaced by any other object type. As mentioned in the threads discussion - I would also think that it would be no problem if you restrict the NodeRed Adapter to create objects under only under 0_userdata - but must not follow rules as which would be necessary for other adapter name spaces.

**Comment by Apollon77** [comment]
Please check Github

**Comment by mickym2** [comment]
There is no difference - no folder object was created in the new tree.

![image](https://user-images.githubusercontent.com/58858526/157857984-f1cb8159-dbc6-4a00-b325-4e9ae4475c02.png)

**Comment by Apollon77** [comment]
Hm ... ok ... warte ... aktuell wurdd es nur gemacht wenn ein neues Objekt angelegt wurde ... GitHub ist aktualisiert ... dann wirds immer geprüft. Am bersten vor restart debug log einschalten , dann sollte man es auch sehen

**Comment by mickym2** [comment]
Hmm - ich sehe zwar das an einer anderen Stelle, wohl ein fehlendes Objekt erstellt wurde

```
2022-03-11 13:48:13.443 - debug: node-red.0 (23978) Create folder object for 0_userdata.0.stromverbrauch

2022-03-11 13:48:13.465 - debug: node-red.0 (23978) Create folder object for 0_userdata.0.dimmer

2022-03-11 13:48:13.468 - debug: node-red.0 (23978) Create folder object for 0_userdata.0.stromverbrauch.shellies

2022-03-11 13:48:13.757 - debug: node-red.0 (23978) Create folder object for 0_userdata.0.stromverbrauch.shellies.shellyplug-s-01E405_0
```

Aber der Großteil bei den Neuanlagen wird nur der state erstellt.
[NR Debug Log.txt](https://github.com/ioBroker/ioBroker.node-red/files/8232017/NR.Debug.Log.txt)

Ich hab das Log hier mal angehängt - da es wenn man den Text hier reinkopiert auch in CodeTags komisch aussieht.

**Comment by mickym2** [comment]
Leider keine Veränderung. 
[iobroker.current.log](https://github.com/ioBroker/ioBroker.node-red/files/8232785/iobroker.current.log)

**Comment by Apollon77** [comment]
Wtf .. ;-)I added debug ... hopefully we know more afterwards

**Comment by Apollon77** [comment]
Klick im Admin mal reload bitte ... laut ihm sind die objekte da :-))

Ich hab nochmal was optimizert das nicht zuvieles mehrfach versucht wird ... von daher bitte nochmal checken ...oder mal baum "Löschen" und ganz neu anlegen lassen. Noch ein letztes Debug log wäre cool

**Comment by mickym2** [comment]
Es tut mir leid - auch wenn das Log was anderes behauptet. Die Objekte sind nicht da!!!

Habe Browsercache gelöscht, alten Objektbaum gelöscht, Browser zu und auf. An dem liegt es nicht.
[CurrentLog.txt](https://github.com/ioBroker/ioBroker.node-red/files/8234583/CurrentLog.txt)

![image](https://user-images.githubusercontent.com/58858526/157932468-66df7ed3-316e-4289-a42c-916ac563e05a.png)

**Comment by Apollon77** [comment]
Also das letzte Log zeigt das node-red nicht neu gestartet ist nach dem GitHub update! ... scheinbar wurde nur der Flow neu laufen gelassen? Kannste bitte mal adapter restarten?

**Comment by mickym2** [comment]
Selbes Bild wie vorher - keine Objekte vorhanden.
[CurrentLog.txt](https://github.com/ioBroker/ioBroker.node-red/files/8234689/CurrentLog.txt)

Adapterneustart: 2022-03-11 19:55:19.734

**Comment by Apollon77** [comment]
Also jetzt blicke ich nix mehr 

Mach mal bitte an einer Shell:

`iob object get 0_userdata.0.mwReImport.mqtt.1.zigbee2mqtt.temperatur.kueche`

**Comment by mickym2** [comment]
`pi@MWHome:~ $ iob object get 0_userdata.0.mwReImport.mqtt.1.zigbee2mqtt.temperatur.kueche

The object 0_userdata.0.mwReImport.mqtt.1.zigbee2mqtt.temperatur.kueche was not found!`

Hier kann man das schlechter als im Forum lesen deshalb nochmal der Screenshot

![image](https://user-images.githubusercontent.com/58858526/157936128-d0ad87ce-74a8-4b9d-8f4b-a61d765e09c7.png)

**Comment by Apollon77** [comment]
Please another github update ... think found the issue ... make sure it restarts please

**Comment by mickym2** [comment]
Perfect - 100% done. :) - 

![image](https://user-images.githubusercontent.com/58858526/157947038-6616c777-0709-4bf9-a274-d8e737fa0948.png)

[CurrentLog.txt](https://github.com/ioBroker/ioBroker.node-red/files/8235080/CurrentLog.txt)

Auch das nachträgliche Umwandeln eines Folder Objektes in einen State funktioniert wieder. Klasse!!!

**Comment by Apollon77** [comment]
Released in 3.0.0
